src/messaging/ExchangeMessageDispatch.cpp

```diff
@@ -94,10 +94,15 @@ CHIP_ERROR ExchangeMessageDispatch::SendMessage(SecureSessionHandle session, uin
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR ExchangeMessageDispatch::OnMessageReceived(const PayloadHeader & payloadHeader, uint32_t messageId,
-                                                      const Transport::PeerAddress & peerAddress, MessageFlags msgFlags,
-                                                      ReliableMessageContext * reliableMessageContext)
+CHIP_ERROR ExchangeMessageDispatch::OnMessageReceived(const Header::Flags & headerFlags, const PayloadHeader & payloadHeader,
+                                                      uint32_t messageId, const Transport::PeerAddress & peerAddress,
+                                                      MessageFlags msgFlags, ReliableMessageContext * reliableMessageContext)
 {
+    if (IsEncryptionRequired())
+    {
+        VerifyOrReturnError(headerFlags.Has(Header::FlagValues::kEncryptedMessage), CHIP_ERROR_INVALID_ARGUMENT);
+    }
```

Comment:
I think, based on spec, where unencrypted messages are received, we need to check that the SessionID is 0 in the message.

Overall, it should have never happened that an unencrypted message is seen in an exchange that requires encryption because:

1. An exchange is bound to a session. An exchange ID cannot cross sessions
2. The session metadata associated with a given exchange ID, if an unencrypted message is received, should have indicated (through the fact that session ID is 0 or non-zero) whether the exchange is for a secure session, and the session type should allow it as well.

In other words, I am not 100% sure that setting that "IsEncryptedRequired()" on the specific set of "SessionEstablishment" message exchange dispatch is the general solution.

@yufengwangca @bzbarsky-apple ?

Reply:
@tcarmelveilleux is session ID 0 reserved for unencrypted messages? Or, can an encrypted message be sent using session ID 0? 
Actually, I could use some help finding the section in the spec that defines it (I tried looking real quick but could not find the right section).

If the sender is a malicious node, the unencrypted message could be delivered to an exchange that requires encryption. Actually, @yufengwangca pointed this out last week, that triggered this change. Maybe the spec already prevents it through some other mechanism, and I can update this PR to follow that approach.

Reply:
Here are the parts that I "followed" in my thinking:

![image](https://user-images.githubusercontent.com/1313847/126681219-294c0e41-84d9-43d3-b324-8c2a36eb5b23.png)

^^^ Encryption type != 0 --> must have a session ID != 0. But conversely, encryption type == 0, must not have a session ID

If you don't have a session ID, then you are going through the unencrypted path altogether. The message is not allowed to have Encryption Type == 0 and a session ID. 

However, reading the spec, I see that the current state does not enforce "one exchange must be mapped to a given session" (e.g. can't cross sessions for a same exchange).

What I thought was already present, though, is the "exchanges are associated with a session". That's a yet-to-be-merged PR: https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/3588 and there is significant discussion there.

Overall, I think that since most of the underlying protocol assumptions of incoming Weave messaging protocol have been changed over time, I think we are left with several different inconsistencies that are not well understood uniformly by either the SDK team, or the spec team. This is a big problem and we should talk about it.

Given the current (and in my opinion, internally inconsistent) state of the spec, your PR solves the problem, but it ties together concepts that are not quite tied that way in the spec. I would argue that we need to get to a point where it's totally unambiguous how the following relate:

- Session IDs and counter spaces in messages
- Exchanges and sessions for both unencrypted (session establishment) and encrypted (CASE/PASE sessions) communication
- How RMP can use enencrypted exchange space cleanly
- etc

Reply:
@msandstedt @turon @bzbarsky-apple PTAL ^^^

Reply:
> Given the current (and in my opinion, internally inconsistent) state of the spec, your PR solves the problem, but it ties together concepts that are not quite tied that way in the spec. I would argue that we need to get to a point where it's totally unambiguous how the following relate:

@tcarmelveilleux while the spec is being updated what should we do about this PR?

1. Close the PR and file an SDK issue to track this problem
2. Merge the PR, and file an SDK issue to revisit it when the spec is updated

The change is not huge, and can easily be updated to match the spec when the spec changes are decided. Given this is a potential security problem, I would like to err on the side of caution..

Also, the unit test in the PR _would_ be useful whenever we update the code to match the spec.

Reply:
> Encryption type != 0 --> must have a session ID != 0

Where does the spec says that?   If encryption type is != 0, then a session id field must be present.  The high bit must be 0 for CASE/PASE and 1 for group.  The other bits can be anything at all, no?  @tcarmelveilleux @turon 

> conversely, encryption type == 0, must not have a session ID

That's correct.  No session id at all in the message...

But apart from all this, what is the proposed logic flow here?  

To be clear, the main thing this PR is aiming to fix is this situation:

1. We have received a message.
2. This is the first message in the exchange (we are not the initiator, the other side is).
3. The message has encryption type set to 0.
4. Our unsolicited message handler for the message protocol/opcode only wants to get encrypted messages.

What this is aiming for is to add a mechanism whereby a UMH can declare that it only wants to get encrypted messages and a mechanism for enforcing such a declaration.

Reply:
I think from that perspective, I agree with you, and can approve this PR on that basis of addressing that subset problem.

Reply:
> while the spec is being updated what should we do about this PR?
> 
>     1. Close the PR and file an SDK issue to track this problem
> 
>     2. Merge the PR, and file an SDK issue to revisit it when the spec is updated
> 
> 
> The change is not huge, and can easily be updated to match the spec when the spec changes are decided. Given this is a potential security problem, I would like to err on the side of caution..

> Given this is a potential security problem, I would like to err on the side of caution..

The problem is not that we're accepting unencrypted messages when encrypted messages were expected.  The problem is that we aren't matching exchanges to specific sessions.  This ceases to be a problem if we simply bind each exchange to a single session.

And setting this flag doesn't entirely fix the security problem.  This still leaves the possibility of another actor instantiating a different secure session, and then hijacking the exchange ID.  That's privilege escalation: still extremely serious.

Personally, I think this can be resolved simply and cleanly in the spec and I don't see a compelling reason to introduce a non-spec-derived patch that only partially solves the security problem.  I think general understanding of the session IDs is not widespread and I'm concerned that introducing more non-spec code paths may further confuse things.

Reply:
Taking this all together, isn't this the logic we want?

```
if (session ID not enclosed and encryption type == 0) {
  invoke unsecured message handler(s), i.e. handlers associated with null session
} else if (session ID enclosed and encryption type == 0) {
  discard
} else if (session ID enclosed and session cannot be located) {
  discard
} else if (session ID enclosed and MIC shows decryption failure) {
  discard
} else {
  pass to session-specific context
}
```

In this case, session-specific context is an exchange context.  But the above inverts responsibilities from what is proposed in this PR.  The exchange context should not be opting into or out of encryption.  Rather, the session layer should never forward something to a context that does not belong to that context.

In contrast, this PR is telling me that we can somehow get to the bottom of this if/else statement and still need an encryption required / not required flag to do the right thing.  If that is so, it feels like security in the code is very broken.  I don't see how this PR actually addresses that.

Reply:
@msandstedt In this scenario, when unencrypted message is received (i.e. session ID is not enclosed, and encryption type is 0), it's handed over to the UMH, as the exchange does not exist. Currently we do not differentiate between UMHs based on whether it can accept unencrypted messages (with no session ID, enc type 0), or it can only accept encrypted messages (with valid session ID, and enc type != 0). This PR is trying to add this differentiation. Maybe the arguments, and variables can be renamed to align more with the proposed spec changes.

So, would this resolve the issue if instead of checking if a UMH can accept unencrypted messages, we check if the UMH can accept a message received with <no session ID, and enc type 0>?

We still need to find a way to bind an exchange (or a collection of exchanges) to a secure session ID. That functionality can be added on top of the proposed update to this PR, when the spec has more clarity.

Reply:
@pan-apple ,

Thank you for this explanation.  This is the piece I was missing:

> In this scenario, when unencrypted message is received (i.e. session ID is not enclosed, and encryption type is 0), it's handed over to the UMH, as the exchange does not exist.

I had misunderstood this entry point as only feeding into existing exchanges.  But obviously for unsolicited messages, something must trigger creation of the exchange context.  This is apparently the call chain where that happens.  So this makes sense.

> So, would this resolve the issue if instead of checking if a UMH can accept unencrypted messages, we check if the UMH can accept a message received with <no session ID, and enc type 0>?

No, this abstraction is fine.  I assume the session layer should discard unencrypted messages with session ID present, but that's probably a session layer function.

My remaining concerns are around CHIP-Specifications/connectedhomeip-spec#3588.  It's clear to me now that this is separate.